src/Compilers/CSharp/Portable/Lowering/LocalRewriter/DelegateCacheRewriter.cs

```diff
@@ -0,0 +1,180 @@
+
+using System.Collections.Generic;
+using System.Diagnostics;
+using Microsoft.CodeAnalysis.CSharp.Symbols;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.CSharp;
+
+/// <summary>
+/// This type helps rewrite the delegate creations that target static method groups to use a cached instance of delegate.
+/// </summary>
+internal sealed partial class DelegateCreationRewriter
+{
+    private readonly SyntheticBoundNodeFactory _factory;
+    private readonly int _topLevelMethodOrdinal;
+
+    private Dictionary<MethodSymbol, DelegateCacheContainer>? _genericCacheContainers;
+
+    internal DelegateCreationRewriter(SyntheticBoundNodeFactory factory, int topLevelMethodOrdinal)
+    {
+        Debug.Assert(factory.TopLevelMethod is { });
+
+        _factory = factory;
+        _topLevelMethodOrdinal = topLevelMethodOrdinal;
+    }
+
+    internal static bool CanRewrite(MethodSymbol targetMethod, BoundConversion boundConversion)
+        => targetMethod.IsStatic && !boundConversion.IsExtensionMethod;
+
+    internal BoundExpression Rewrite(BoundDelegateCreationExpression boundDelegateCreation, MethodSymbol targetMethod, TypeSymbol delegateType)
+    {
+        Debug.Assert(delegateType.IsDelegateType());
+
+        var oldSyntax = _factory.Syntax;
+        _factory.Syntax = boundDelegateCreation.Syntax;
+
+        var cacheContainer = GetOrAddCacheContainer(delegateType, targetMethod);
+        var cacheField = cacheContainer.GetOrAddCacheField(_factory, delegateType, targetMethod);
+
+        var boundCacheField = _factory.Field(null, cacheField);
+        var rewrittenNode = _factory.Coalesce(boundCacheField, _factory.AssignmentExpression(boundCacheField, boundDelegateCreation));
+
+        _factory.Syntax = oldSyntax;
+
+        return rewrittenNode;
+    }
+
+    private DelegateCacheContainer GetOrAddCacheContainer(TypeSymbol delegateType, MethodSymbol targetMethod)
+    {
+        Debug.Assert(_factory.ModuleBuilderOpt is { });
+        Debug.Assert(_factory.CurrentFunction is { });
+
+        var typeCompilationState = _factory.CompilationState;
+        var generation = _factory.ModuleBuilderOpt.CurrentGenerationOrdinal;
+
+        DelegateCacheContainer? container;
+
+        if (CanUseTypeScopedConcreteCacheContainer(delegateType, targetMethod))
+        {
+            container = typeCompilationState.ConcreteDelegateCacheContainer;
+
+            if (container is { })
+            {
+                return container;
+            }
+
+            container = new DelegateCacheContainer(typeCompilationState.Type, generation);
+            typeCompilationState.ConcreteDelegateCacheContainer = container;
+        }
+        else
+        {
+            // We don't need to synthesize a container for each and every function.
+            // For functions with zero arity, we can share the container with the closest generic ancestor function.
```

Comment:
>For functions with zero arity, we can share the container with the closest generic ancestor function. [](http://example.com/codeflow?start=15&length=101)

I would like to suggest an additional improvement. Right now we collect generic type parameters of enclosing methods and then check if any of them are used by the delegate or the target method. We can flip the genericity check around. We can collect type parameters used by the delegate and the target method instead (using the same ```VisitType``` helper that ```ContainsTypeParameters``` uses). Then we go up the chain of containing methods checking if any of their type parameters are among the set we collected. Once we hit a match, that is the owner for the cache container. This way we will be able to skip some generic methods as well. If we got to the top, the type level cache container should be used. Local functions as targets will have to be handled specially, for those we just stop at the first generic containing method.

Reply:
Good idea, I'll try this out.

Reply:
Implemented. Please check.